### PR TITLE
Rewrite descriptor and digest cache

### PR DESCRIPTION
**Pull request by legionus**
* Add a new caching subsystem;
* The descriptor cache is integrated with the digest cache;
* Descriptors have a TTL;

### CODE REVIEW

File: pkg/dockerregistry/server/repository.go [hunk 1]
```diff
@@ -58,12 +59,10 @@ type repository struct {
 	cachedImages map[digest.Digest]*imageapiv1.Image
 	// cachedImageStream stays cached for the entire time of handling signle repository-scoped request.
 	imageStreamGetter *cachedImageStreamGetter
-	// cachedLayers remembers a mapping of layer digest to repositories recently seen with that image to avoid
-	// having to check every potential upstream repository when a blob request is made. The cache is useful only
-	// when session affinity is on for the registry, but in practice the first pull will fill the cache.
-	cachedLayers digestToRepositoryCache
 	// remoteBlobGetter is used to fetch blobs from remote registries if pullthrough is enabled.
 	remoteBlobGetter BlobGetterService
+	// cache TODO
```
Comment (dmage) [inline pkg/dockerregistry/server/repository.go]: TODO

File: pkg/dockerregistry/server/repository.go [hunk 2]
```diff
@@ -328,9 +334,19 @@ func (r *repository) rememberLayersOfImage(image *imageapiv1.Image, cacheName st
...
+
 	if len(image.DockerImageLayers) > 0 {
 		for _, layer := range image.DockerImageLayers {
-			r.cachedLayers.RememberDigest(digest.Digest(layer.Name), r.app.config.Cache.BlobRepositoryTTL, cacheName)
+			_ = descCache.SetDescriptor(r.ctx, digest.Digest(layer.Name), distribution.Descriptor{
+				Digest:    digest.Digest(layer.Name),
+				Size:      layer.LayerSize,
+				MediaType: layer.MediaType,
+			})
+			//_ = r.cache.AddDigest(digest.Digest(layer.Name), cacheName)
```
Comment (dmage) [inline pkg/dockerregistry/server/repository.go]: Remove it?

**Comment by legionus** [comment]
> Don't you want to disable the upstream cache?
@dmage done

**Pull request merged**